File: omnivault-frontend/src/services/tagService.js
import axiosInstance from "./axiosInstance";
import { apiCache } from "../utils/apiCache";

const tagService = {
  getAllTags: async () => {
    const response = await axiosInstance.get("/tags");
    return response.data;
  },

  getTag: async (tagId) => {
    const cacheKey = `tag_${tagId}`;

    // Try to get from cache first
    const cachedTag = apiCache.get(cacheKey);
    if (cachedTag) {
      return cachedTag;
    }

    // Not in cache, fetch from API
    const response = await axiosInstance.get(`/tags/${tagId}`);

    // Store in cache before returning
    apiCache.set(cacheKey, response.data);
    return response.data;
  },

  createTag: async (tagData) => {
    const response = await axiosInstance.post("/tags", tagData);
    return response.data;
  },

  updateTag: async (tagId, tagData) => {
    const response = await axiosInstance.put(`/tags/${tagId}`, tagData);
    return response.data;
  },

  deleteTag: async (tagId) => {
    await axiosInstance.delete(`/tags/${tagId}`);
  },

  searchTags: async (query) => {
    const response = await axiosInstance.get(`/tags/search?query=${query}`);
    return response.data;
  },
};

export default tagService;

File: omnivault-backend/src/main/java/com/omnivault/service/TagService.java
package com.omnivault.service;

import com.omnivault.domain.dto.request.TagCreateRequest;
import com.omnivault.domain.dto.response.TagDTO;
import com.omnivault.domain.model.Tag;

import java.util.List;
import java.util.Set;
import java.util.UUID;

/**
 * Service for managing user tags.
 * Provides comprehensive tag management operations including
 * creation, retrieval, update, and deletion of tags. Supports
 * advanced tag-related functionalities like searching and
 * bulk tag operations.
 */
public interface TagService {

    /**
     * Get all tags for the current user
     *
     * @return List of tagsa
     */
    List<TagDTO> getAllTags();

    /**
     * Get tag by ID
     *
     * @param tagId The tag ID
     * @return The tag DTO
     */
    TagDTO getTag(UUID tagId);

    /**
     * Get tag entity by ID
     *
     * @param tagId The tag ID
     * @return The tag entity
     */
    Tag getTagEntity(UUID tagId);

    /**
     * Create a new tag
     *
     * @param request The tag creation request
     * @return The created tag DTO
     */
    TagDTO createTag(TagCreateRequest request);

    /**
     * Update a tag
     *
     * @param tagId The tag ID
     * @param request The tag update request
     * @return The updated tag DTO
     */
    TagDTO updateTag(UUID tagId, TagCreateRequest request);

    /**
     * Delete a tag
     *
     * @param tagId The tag ID
     */
    void deleteTag(UUID tagId);

    /**
     * Find or create tags by names
     *
     * @param tagNames List of tag names
     * @return Set of tag entities
     */
    Set<Tag> findOrCreateTags(List<String> tagNames);

    /**
     * Get tags by IDs
     *
     * @param tagIds List of tag IDs
     * @return Set of tag entities
     */
    Set<Tag> getTagsByIds(List<UUID> tagIds);

    /**
     * Search tags by name
     *
     * @param searchTerm The search term
     * @return List of matching tags
     */
    List<TagDTO> searchTags(String searchTerm);
}